cpp/SnowflakeS3Client.cpp

```diff
@@ -299,29 +308,193 @@ void SnowflakeS3Client::addUserMetadata(
   userMetadata->insert({SFC_DIGEST, fileMetadata->sha256Digest});
 }
 
-void SnowflakeS3Client::extractBucketAndKey(FileMetadata *fileMetadata,
+void SnowflakeS3Client::extractBucketAndKey(std::string *fileFullPath,
                                             std::string &bucket,
                                             std::string &key)
 {
-  size_t sepIndex = m_stageInfo->getLocation()->find_first_of('/');
-  bucket = m_stageInfo->getLocation()->substr(0, sepIndex);
-  key = m_stageInfo->getLocation()->substr(sepIndex + 1)
-        + fileMetadata->destFileName;
+  size_t sepIndex = fileFullPath->find_first_of('/');
+  bucket = fileFullPath->substr(0, sepIndex);
+  key = fileFullPath->substr(sepIndex + 1);
 }
 
-TransferOutcome SnowflakeS3Client::handleError(
+RemoteStorageRequestOutcome SnowflakeS3Client::handleError(
   const Aws::Client::AWSError<Aws::S3::S3Errors> & error)
 {
   if (error.GetExceptionName() == "ExpiredToken")
   {
     sf_log_warn(CXX_LOG_NS, "Token expired.");
-    return TransferOutcome::TOKEN_EXPIRED;
+    return RemoteStorageRequestOutcome::TOKEN_EXPIRED;
   }
   else
   {
     sf_log_error(CXX_LOG_NS, "S3 request failed failed: %s",
                  error.GetMessage().c_str());
-    return TransferOutcome::FAILED;
+    return RemoteStorageRequestOutcome::FAILED;
+  }
+}
+
+RemoteStorageRequestOutcome SnowflakeS3Client::download(
+  FileMetadata *fileMetadata,
+  std::basic_iostream<char>* dataStream)
+{
+  if (fileMetadata->srcFileSize > DATA_SIZE_THRESHOLD)
```

Comment:
so what was your solution to get the size before download?

Reply:
FileMetadataInitializer class will issue a aws s3 rest request to get object metadata of the file size. (This is how s3 transfer library did as well)

Reply:
ok